File: server/internal/ports/persistence.go
package ports

import (
	"context"
	"time"

	"github.com/google/uuid"

	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/backtest"
	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/evaluation"
	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/market"
	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/news"
	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/ranking"
	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/search"
)

type CandleReader interface {
	ListClosedCandles(context.Context, market.CandleQuery) ([]market.Candle, error)
}

type CandleWriter interface {
	UpsertClosedCandles(context.Context, []market.Candle) error
}

type DatasetReader interface {
	LoadDatasetCandles(context.Context, uuid.UUID) ([]market.Candle, error)
}

type ExperimentRepository interface {
	CreateSnapshot(context.Context, backtest.ExperimentSnapshot) error
	GetSnapshot(context.Context, uuid.UUID) (backtest.ExperimentSnapshot, error)
}

type RunRepository interface {
	PersistResult(context.Context, uuid.UUID, backtest.Result) error
	OwnsRun(context.Context, uuid.UUID, uuid.UUID) (bool, error)
}

type SearchRunRepository interface {
	Create(context.Context, search.SearchRun) error
	Get(context.Context, uuid.UUID) (search.SearchRun, error)
	ApplyAction(context.Context, uuid.UUID, string, uuid.UUID) error
}

type TradeReader interface {
	ListTrades(context.Context, uuid.UUID, int, int) ([]backtest.TradeFact, error)
}

type EquityReader interface {
	ListEquity(context.Context, uuid.UUID, int) ([]backtest.EquityPoint, error)
}

type EvaluationReader interface {
	GetEvaluation(context.Context, uuid.UUID, string) (evaluation.Evaluation, error)
}

type LeaderboardRepository interface {
	List(context.Context, string, string, int) ([]ranking.LeaderboardEntry, error)
	Insert(context.Context, ranking.LeaderboardEntry) error
}

type NewsRepository interface {
	ListSources(context.Context) ([]news.ApprovedSource, error)
	InsertItems(context.Context, []news.Item) ([]news.Item, error)
}

type EventStore interface {
	Append(context.Context, []byte) error
	ClaimForConsumer(context.Context, uuid.UUID, string) (bool, error)
}

type TransactionBoundary interface {
	WithinTransaction(context.Context, func(context.Context) error) error
}

type ReadinessChecker interface {
	Ready(context.Context) error
}

type Clock interface{ Now() time.Time }

File: py_backend/app/ports/persistence.py
"""Persistence seams for the moved domain. Mirrors `server/internal/ports/persistence.go` subset."""

from __future__ import annotations

from typing import Protocol
from uuid import UUID

from ..domain.backtest import EquityPoint, ExperimentSnapshot, Result, TradeFact
from ..domain.evaluation import Evaluation
from ..domain.ranking import LeaderboardEntry
from ..domain.search import SearchRun


class DatasetReader(Protocol):
    def load_dataset_candles(self, dataset_id: UUID) -> list: ...


class ExperimentRepository(Protocol):
    def create_snapshot(self, snapshot: ExperimentSnapshot) -> None: ...

    def get_snapshot(self, experiment_id: UUID) -> ExperimentSnapshot: ...


class RunRepository(Protocol):
    def persist_result(self, run_id: UUID, result: Result) -> None: ...

    def owns_run(self, run_id: UUID, owner_id: UUID) -> bool: ...


class SearchRunRepository(Protocol):
    def create(self, run: SearchRun) -> None: ...

    def get(self, run_id: UUID) -> SearchRun: ...

    def apply_action(self, run_id: UUID, action: str, operator_id: UUID) -> None: ...


class TradeReader(Protocol):
    def list_trades(self, run_id: UUID, limit: int, offset: int) -> list[TradeFact]: ...


class EquityReader(Protocol):
    def list_equity(self, run_id: UUID, max_points: int) -> list[EquityPoint]: ...


class EvaluationReader(Protocol):
    def get_evaluation(self, run_id: UUID, version: str) -> Evaluation: ...


class LeaderboardRepository(Protocol):
    def list(self, timeframe: str, symbol: str, top_k: int) -> list[LeaderboardEntry]: ...

    def insert(self, entry: LeaderboardEntry) -> None: ...
